library/psa_crypto.c

```diff
@@ -1260,6 +1260,137 @@ static psa_status_t psa_validate_optional_attributes(
+    if (key_bits_attr == 0) {
+        return PSA_SUCCESS;
+    }
+
+    /* Check if the size matches the curve family */
+    switch (PSA_KEY_TYPE_ECC_GET_FAMILY(type)) {
+        case PSA_ECC_FAMILY_SECP_K1:
+            if (key_bits_attr != 192 && key_bits_attr != 225 && key_bits_attr != 256) {
+                return PSA_ERROR_INVALID_ARGUMENT;
+            }
+            break;
+        case PSA_ECC_FAMILY_SECP_R1:
+            if (key_bits_attr != 192 && key_bits_attr != 224 && key_bits_attr != 256 &&
+                key_bits_attr != 384 && key_bits_attr != 521) {
+                return PSA_ERROR_INVALID_ARGUMENT;
+            }
+            break;
+        case PSA_ECC_FAMILY_SECT_K1:
+            if (key_bits_attr != 233 && key_bits_attr != 239 && key_bits_attr != 283 &&
+                key_bits_attr != 409 && key_bits_attr != 571) {
+                return PSA_ERROR_INVALID_ARGUMENT;
+            }
+            break;
+        case PSA_ECC_FAMILY_SECT_R1:
+            if (key_bits_attr != 233 && key_bits_attr != 283 && key_bits_attr != 409 &&
+                key_bits_attr != 571) {
+                return PSA_ERROR_INVALID_ARGUMENT;
+            }
+            break;
+        case PSA_ECC_FAMILY_BRAINPOOL_P_R1:
+            if (key_bits_attr != 192 && key_bits_attr != 224 && key_bits_attr != 256 &&
+                key_bits_attr != 320 && key_bits_attr != 384 && key_bits_attr != 512) {
+                return PSA_ERROR_INVALID_ARGUMENT;
+            }
+            break;
+        case PSA_ECC_FAMILY_TWISTED_EDWARDS:
+        case PSA_ECC_FAMILY_MONTGOMERY:
+            if (key_bits_attr != 255 && key_bits_attr != 448) {
+                return PSA_ERROR_INVALID_ARGUMENT;
+            }
+            break;
+        default:
+            return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    return PSA_SUCCESS;
+}
+
+static psa_status_t psa_validate_ecc_key_data_length(const psa_key_attributes_t *attributes,
+                                                     size_t data_length)
+{
+    psa_algorithm_t alg = psa_get_key_algorithm(attributes);
+    psa_key_type_t type = psa_get_key_type(attributes);
+    size_t key_data_length_bits = PSA_BYTES_TO_BITS(data_length);
+
+    /* With algorithm we can't evaluate more fields */
+    if (alg == PSA_ALG_NONE) {
+        return PSA_SUCCESS;
+    }
+
+    /* Check if the size matches the curve family */
+    switch (PSA_KEY_TYPE_ECC_GET_FAMILY(type)) {
+        case PSA_ECC_FAMILY_SECP_R1:
+            /* secpr1p521 will require 66 bytes therefore checking for 528 bits */
+            if (key_data_length_bits != 192 && key_data_length_bits != 224 &&
+                key_data_length_bits != 256 && key_data_length_bits != 384 &&
+                key_data_length_bits != 528) {
+                return PSA_ERROR_INVALID_ARGUMENT;
+            }
+            break;
+        case PSA_ECC_FAMILY_SECT_K1:
+            if (key_data_length_bits != 233 && key_data_length_bits != 239 &&
```

Comment:
Some of the values that key_data_length_bits are checked against are not possible to get from the PSA_BYTES_TO_BITS(data_length).

Reply:
Yeah changed everything to bytes